### ISSUE TITLE
System.MissingMethodException

### ISSUE THREAD

**Issue by brianchance** [jonwagner/Insight.Database#407]
# Describe the bug

We have some long running services that every once in a while start throwing System.MissingMethodException. It believe it cannot find the generated method that used to be there. Restarting the service generally fixes it. In other words, it has been running fine. Maybe memory pressure is unloading the generated code?

I have also seen this during development. If I rebuild all, it works.

Not sure how the IL code is generated, but this one fails on the second call to the same database using the same abstract class. 

```
        public async Task Execute(QueueModel model)
        {
            var order = await _orderAdapter.Load(model.OrderId);
            if (order.SynchedOn == null)
            {
                await _syncAdapter.CreateOrder(order);
                await _orderAdapter.SetSynced(model.OrderId); // THIS LINE THROWS THE System.MissingMethodException
            }
        }
```

Database abstract class
```
    public abstract class OrderAdapter
    {
        public abstract IDbConnection GetConnection();

        private static readonly SingleReader<Order> _orderReader = Query.ReturnsSingle<Order>()
            .ThenChildren(Some<OrderItem>.Records);

        public async Task<Order> Load(int orderId)
        {
            return await GetConnection()
                .QueryAsync("dbo.Order_Get", new
                {
                    orderId
                }, _orderReader);
        }

        [Sql("Order_SetSynced", Schema = "dbo")]
        public abstract Task<string> SetSynced(int orderId);
    }
```

Full dump
```
Inner exception System.MissingMethodException handled at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw:
   at System.ModuleHandle.ResolveMethod (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.ModuleHandle.ResolveMethodHandleInternalCore (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.ModuleHandle.ResolveMethodHandleInternal (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Reflection.CustomAttribute.FilterCustomAttributeRecord (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Reflection.CustomAttribute.GetCustomAttributes (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Reflection.CustomAttribute.GetCustomAttributes (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Insight.Database.CodeGenerator.TypeHelper.IsSqlUserDefinedType (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at Insight.Database.CodeGenerator.TypeHelper.IsAtomicType (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at Insight.Database.CodeGenerator.ClassPropInfo+<>c__DisplayClass52_0.<GetMembersForType>b__0 (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd (System.Collections.Concurrent, Version=4.0.14.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at Insight.Database.CodeGenerator.ClassPropInfo.GetMemberByColumnName (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at Insight.Database.CodeGenerator.ClassPropInfo.SearchForMatchingField (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at Insight.Database.ColumnMapping.MapParameter (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at System.Linq.Enumerable+SelectListIterator`2.ToList (System.Linq, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at System.Linq.Enumerable.ToList (System.Linq, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at Insight.Database.CodeGenerator.DbParameterGenerator.CreateClassInputParameterGenerator (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd (System.Collections.Concurrent, Version=4.0.14.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at Insight.Database.DBCommandExtensions.AddParameters (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at Insight.Database.DBConnectionExtensions.CreateCommand (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at Insight.Database.DBConnectionExtensions+<>c__DisplayClass2_0`1.<ExecuteScalarAsync>b__0 (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at Insight.Database.DBConnectionExtensions+<ExecuteAsyncAndAutoClose>d__34`1.MoveNext (Insight.Database.Core, Version=6.2.9.0, Culture=neutral, PublicKeyToken=f35c6ef87e515afc)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at App.UseCases.Sync.SyncOrder+<Execute>d__3.MoveNext (App, Version=1.0.0.0, Culture=neutral, PublicKeyToken=nullApp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: D:\agent\_work\2\s\App\UseCases\Sync\SyncOrder.csApp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: 28)

```

## Steps to reproduce

I wish I had some to reliably reproduce. 

## Expected behavior

If we trap for the error, can we clear and refresh the generated code? Any help would be appreciated. 

- Dotnet version: [netcore2]
- Database: [SQL Server]
- Library version: [6.2.9]

**Comment by jonwagner** [comment]
Interesting....


  *   Because it's in CreateClassInputParameterGenerator - this is a new query that hasn't been run before (unless there's a thread synchronization issue in insight)
  *   Are you sure the code isn't failing on the await for CreateOrder? That's more likely given the stack trace unless OrderId isn't just an int.
  *   Can you post the code for Order / CreateOrder and the signature for the stored proc it calls?


________________________________

**Comment by brianchance** [comment]
I know it is not failing on CreateOrder, that call runs against another database with a transaction in the sproc and it completes. The SetSynced does not, so when it re-runs we get two orders.

This code has been running without issues for nearly a year. We normally have at least some minor change which deploys and resets the service.

Set Syched
```
CREATE PROCEDURE dbo.Order_SetSynced
	@OrderId INT
```

Order code...
```
public class Order
    {
        public int OrderId { get; set; }
        public string ClientKey { get; set; }
        public string BrandCode { get; set; }
        public string CountryCode { get; set; }

...CUTTING OUT SIMPLE FIELDS FOR BREVITY

        public DateTime? SynchedOn { get; set; }
        public DateTime CreatedOn { get; set; }

        public IEnumerable<OrderItem> Items { get; set; } = new List<OrderItem>();
}

public class OrderItem
    {
        public int OrderItemId { get; set; }
        public int OrderId { get; set; }
        public int CatalogId { get; set; }
        public string ShortCode { get; set; }
        public int CatalogTypeId { get; set; }
        public int? ProductTypeId { get; set; }
        public decimal CatalogPrice { get; set; }

        ...CUTTING OUT SIMPLE FIELDS FOR BREVITY
    }
```

Create_Order
```
CREATE PROCEDURE dbo.Order_Create
  @CartId INT
  , @ClientKey UNIQUEIDENTIFIER
  , @BrandCode VARCHAR(3)
  , @CountryCode VARCHAR(2)

 ...CUTTING OUT SIMPLE FIELDS FOR BREVITY

  , @Items dbo.CartItemTable READONLY

```

**Comment by jonwagner** [comment]
Interesting. Then maybe there is an IL stability issue with .net core? #408 just popped up with a similar issue.

What version of .net are you using?

On Nov 1, 2019, at 4:32 PM, brianchance <notifications@github.com> wrote:

﻿

I know it is not failing on CreateOrder, that call runs against another database with a transaction in the sproc and it completes. The SetSynced does not, so when it re-runs we get two orders.

This code has been running without issues for nearly a year. We normally have at least some minor change which deploys and resets the service.

Set Syched

CREATE PROCEDURE dbo.Order_SetSynced
        @OrderId INT


Order code...

public class Order
    {
        public int OrderId { get; set; }
        public string ClientKey { get; set; }
        public string BrandCode { get; set; }
        public string CountryCode { get; set; }

...CUTTING OUT SIMPLE FIELDS FOR BREVITY

        public DateTime? SynchedOn { get; set; }
        public DateTime CreatedOn { get; set; }

        public IEnumerable<OrderItem> Items { get; set; } = new List<OrderItem>();
}

public class OrderItem
    {
        public int OrderItemId { get; set; }
        public int OrderId { get; set; }
        public int CatalogId { get; set; }
        public string ShortCode { get; set; }
        public int CatalogTypeId { get; set; }
        public int? ProductTypeId { get; set; }
        public decimal CatalogPrice { get; set; }

        ...CUTTING OUT SIMPLE FIELDS FOR BREVITY
    }


Create_Order

CREATE PROCEDURE dbo.Order_Create
  @CartId INT
  , @ClientKey UNIQUEIDENTIFIER
  , @BrandCode VARCHAR(3)
  , @CountryCode VARCHAR(2)

 ...CUTTING OUT SIMPLE FIELDS FOR BREVITY

  , @Items dbo.CartItemTable READONLY

**Comment by brianchance** [comment]
we are using dotnet core 2.1

**Comment by resnikb** [comment]
@brianchance Do you happen to have the original exception message, and the full signature of the missing method?

**Comment by brianchance** [comment]
@resnikb - the missing method was SetSynced, the rest of the dump is just an Azure function app wrapper.

**Comment by jonwagner** [comment]
@resnikb identified an area where the interface generator code was not thread-safe. this should be fixed in v6.2.11.

**Comment by brianchance** [comment]
Awesome, thank you!
